### ISSUE TITLE
Results of iCaRL

### ISSUE THREAD

**Issue by qsunyuan** [mmasana/FACIL#12]
Hi! 

Nice work on continual learning.

Recently I reproduced iCaRL, but I cannot achieve the original results. I tried many methods, adjust learning rate, more training epochs, different weight decay.....

And I found your results are quite similar to mine. 

Could u pls give some insight about the original results.

In addition, in BiC, their results are about 50% (the 10th task top-1 test acc), this is very surprising, am I missing something?

Hope to get ur replay. This has bothered me for a long time.

--------------------------------------------------------------------------------------------------
From iCaRL <https://arxiv.org/pdf/1611.07725.pdf>

<img width="408" alt="1641880699(1)" src="https://user-images.githubusercontent.com/71776283/148889306-93555cef-49aa-4d5b-a928-23db1d40f5da.png">

--------------------------------------------------------------------------------------------------
From yours <https://arxiv.org/pdf/2010.15277.pdf>

<img width="612" alt="1641879443(1)" src="https://user-images.githubusercontent.com/71776283/148887237-22fcc002-0d6f-4d1e-916c-3559c77894d3.png">

--------------------------------------------------------------------------------------------------
From BiC <https://arxiv.org/pdf/1905.13260.pdf>

<img width="243" alt="1641881054(1)" src="https://user-images.githubusercontent.com/71776283/148889933-44e2e11e-337b-406b-bec7-71d573fe12c7.png">
<img width="62" alt="1641881167(1)" src="https://user-images.githubusercontent.com/71776283/148890031-4bdcf232-dd18-4da8-8ec9-d9656d9ed081.png">

**Comment by qsunyuan** [comment]
This question may similar to this link.

https://github.com/mmasana/FACIL/issues/6

**Comment by mmasana** [comment]
Hi @qsunyuan,

indeed, the comments from issue #6 are part of the explanation. As far as I remember, another reason is that when iCaRL came out there was not much discussion yet about how evaluation on the different outputs for each task should be done. Make sure to check if some of the results are given under the task-IL setting (when you have the task-ID at inference time). Also, I think in their original implementation, the heads were all created from the beginning and accessed by the optimizer, which technically is information that you should not know in advance. Further, hyperparameters and training regimes were chosen to obtain the best results on that setting at the end of all tasks. Quite a few of those details quickly pile up, and I've talked with some other people that have not managed to reproduce the results in the iCaRL paper for class-IL. I would look at iCaRL more to understand the ideas that they proposed and why, but at this point in time, almost every single incremental method has been able to surpass it when using more realistic continual learning scenarios. Finally, the results we provide in the survey (the figure that you posted above) is an average of 10 runs using the CHL (Continual Hyperparameter Framework), and thus it cannot be compared to the other graphs since it becomes more challenging.

Regarding BiC, I contacted the author some time ago when we released the code and he confirmed that it looked correct. If I remember correctly, we managed to reproduce their results when using the training parameters and regimes from their paper. However, when using the CHF (more realistic) their results go a bit down (as do for all methods).

Hope this helps!

**Comment by qsunyuan** [comment]
Thx for ur help, it really helps a lot. 

Have a good day.
